### PR TITLE
feat: refactor peer class

### LINKED ISSUE #54: Outgoing peer connections - peer protocols
See https://github.com/libp2p/js-libp2p-pubsub/blob/master/src/index.js#L190

It appears that when we connect to a new peer, we attach all protocol strings that _we_ currently support to that peer, rather than only attaching the protocol for the stream that gets opened, as is done in `_onIncomingStream`. Is this intended?

### PR DESCRIPTION
**Pull request by wemeetagain**
Resolves #54 and #56
- `Peer` has been renamed/replaced with `PeerStreams` class that manages incoming and outgoing streams.
- `_addPeer` and `_removePeer` have been slightly modified to suit - _removePeer closes a peer's streams
- `getSubscribers` now uses `this.topics` rather than `peer.topics` to determine the subscribers for a topic

`PeerStreams` manages:
- `id` - the `PeerId` of the peer
- `protocol` - the protocol negotiated with the peer (previously was an array of protocols, which is described in #54 )
- `_rawOutboundStream` (previously named `this.conn`) - this is the stream returned from `conn.newStream`
- `outboundStream` (previously named `this.stream`) - this is the pushable stream to write outgoing messages to. Data pushed here is sent to _rawOutboundStream after first being transformed with a length-prefix encoding transform.
It now manages an inbound stream as well
- `_rawInboundStream` - this is the stream returned in the callback attached to `registrar.handle`. Called when a peer dials us.
- `inboundStream` - this is an abortable stream with first a length-prefix decoding tranform applied. This is the stream that should be read from to receive incoming rpc messages.
- `async attachOutboundStream(stream)` - method to attach a new outgoing stream for management. If the stream is the first outgoing stream to be attached, this method triggers a `stream:outbound` event. If there is already an existing outgoing stream, the existing stream is gently shutdown (without triggering a `close` event) first, then the new stream is attached.
- `attachInboundStream(stream)` - method to attach a new incoming stream for management. If the stream is the first incoming stream to be attached, this method triggers a `stream:inbound` event. If there is already an existing incoming stream, the existing stream is gently shutdown (without triggering a `close` event) first, then the new stream is attached.
- `write(data)` - convenience method to push data to the `outboundStream`, if it exists. If it doesn't, this method throws an error.
- `close()` - method to shutdown both incoming and outgoing streams, triggers a `close` event. In conjunction with the base Pubsub impl, the `close` event will `_removePeer`.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -166,15 +168,16 @@ class PubsubBaseProtocol extends EventEmitter {
    * @private
    * @param {Object} props
    * @param {string} props.protocol
-   * @param {DuplexStream} props.strean
+   * @param {Stream} props.stream
```
Comment (vasco-santos) [inline src/index.js:171]:
```suggestion
   * @param {IterableStream} props.stream
```
Per https://gist.github.com/alanshaw/591dc7dd54e4f99338a347ef568d6ee9, we should call this an IterableStream. This is not a node stream and might create confusion

File: src/index.js [hunk 2]
```diff
@@ -205,49 +207,50 @@ class PubsubBaseProtocol extends EventEmitter {
...
-   * @param {Peer} peer peer state
-   * @returns {Peer}
+   * @param {PeerId} peerId
+   * @returns {PeerStreams | undefined}
    */
-  _removePeer (peer) {
+  _removePeer (peerId) {
+    if (!peerId) return
+    const id = peerId.toB58String()
+    const peer = this.peers.get(id)
```
Comment (vasco-santos) [inline src/index.js:252]:
```suggestion
    const peerStreams = this.peers.get(id)
```

Comment (vasco-santos) [reply src/index.js:252]: Two occurrences below need to change

File: src/peerStreams.js [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+const EventEmitter = require('events')
+
+const lp = require('it-length-prefixed')
+const pushable = require('it-pushable')
+const pipe = require('it-pipe')
+const abortable = require('abortable-iterator')
+const AbortController = require('abort-controller')
+const debug = require('debug')
+
+const log = debug('libp2p-pubsub:peer')
```
Comment (vasco-santos) [inline src/peerStreams.js:12]:
```suggestion
const log = debug('libp2p-pubsub:peer-streams')
```

File: src/peerStreams.js [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+log.error = debug('libp2p-pubsub:peer:error')
```
Comment (vasco-santos) [inline src/peerStreams.js:13]:
```suggestion
log.error = debug('libp2p-pubsub:peer-streams:error')
```

File: src/peerStreams.js [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    this.id = id
+    /**
+     * Established protocol
+     * @type {string}
+     */
+    this.protocol = protocol
+    /**
+     * The raw outbound stream, as retrieved from conn.newStream
+     * @private
+     * @type {Stream}
```
Comment (vasco-santos) [inline src/peerStreams.js:38]: This and the below one should be IterableStream

File: src/peerStreams.js [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+      pipe(
+        this._rawInboundStream,
+        lp.decode()
+      ),
+      this._inboundAbortController.signal,
+      { returnOnAbort: true }
+    )
+
+    // Only emit if the connection is new
+    if (_prevStream) {
```
Comment (vasco-santos) [inline src/peerStreams.js:129]:
```suggestion
    if (!_prevStream) {
```

File: src/peerStreams.js [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+    pipe(
+      this.outboundStream,
+      lp.encode(),
+      this._rawOutboundStream
+    ).catch(err => {
+      log.error(err)
+    })
+
+    // Only emit if the connection is new
+    if (_prevStream) {
```
Comment (vasco-santos) [inline src/peerStreams.js:171]:
```suggestion
    if (!_prevStream) {
```

File: src/index.js (hunk 1 above)
Comment (wemeetagain) [reply src/index.js]: I called it `DuplexIterableStream` so its known to be a duplex object (`{ sink, source }`)

**Pull request merged**